spinalcordtoolbox/deepseg/core.py

```diff
@@ -0,0 +1,144 @@
+        self.param = param
+        self.metadata = metadata
+
+    def threshold(self, nii_seg):
+        """
+        Threshold the prediction. For no prediction, set 'threshold' to 0.
+        """
+        if self.param.threshold:
+            thr = self.param.threshold
+        else:
+            if 'threshold' in self.metadata:
+                thr = self.metadata.threshold
+            else:
+                logger.warning("'threshold' is not defined in the model json file. Using threshold of: {}".format(
+                    DEFAULT_THRESHOLD))
+                thr = DEFAULT_THRESHOLD
+        if thr:
+            nii_seg = imed.postprocessing.threshold_predictions(nii_seg, thr)
+        return nii_seg
+
+    def keep_largest_object(self, nii_seg):
+        """
+        Only keep largest object
+        """
+        # TODO: This if/elif below is ugly. Cannot think of something better for now...
+        do_process = DEFAULT_KEEP_LARGEST_OBJECT
+        if self.param.keep_largest_object is True:
+            do_process = True
+        elif self.param.keep_largest_object is None:
+            if 'keep_largest_object' in self.metadata:
+                do_process = self.metadata.keep_largest_object
+        if do_process:
+            # Make sure input is binary
+            if np.array_equal(nii_seg.get_fdata(), nii_seg.get_fdata().astype(bool)):
+                # Fetch axis corresponding to superior-inferior direction
+                # TODO: move that code in image
+                affine = nii_seg.get_header().get_best_affine()
+                code = nib.orientations.aff2axcodes(affine)
+                if 'I' in code:
+                    axis_infsup = code.index('I')
+                elif 'S' in code:
+                    axis_infsup = code.index('S')
+                else:
+                    raise ValueError(
+                        "Neither I nor S is present in code: {}, for affine matrix: {}".format(code, affine))
+                nii_seg = imed.postprocessing.keep_largest_object_per_slice(nii_seg, axis=axis_infsup)
+            else:
+                logger.warning("Algorithm 'keep largest object' can only be run on binary segmentation. Skipping.")
+        return nii_seg
+
+    def fill_holes(self, nii_seg):
+        """
+        Fill holes
+        """
+        # TODO: This if/elif below is ugly. Cannot think of something better for now...
+        do_process = DEFAULT_FILL_HOLES
+        if self.param.fill_holes is True:
+            do_process = True
+        elif self.param.fill_holes is None:
+            if 'fill_holes' in self.metadata:
+                do_process = self.metadata.fill_holes
+        if do_process:
+            if np.array_equal(nii_seg.get_fdata(), nii_seg.get_fdata().astype(bool)):
+                nii_seg = imed.postprocessing.fill_holes(nii_seg)
+            else:
+                logger.warning("Algorithm 'fill holes' can only be run on binary segmentation. Skipping.")
+            return nii_seg
```

Comment:
This return is tabbed in one level too deep.

Reply:
fixed in bea7a3cb50862cfebe64d36dc71647cf4a7ae1dc